File: server/message_test.go
// Tests for messages parsing/serializing

package main

import "testing"

func compareMessages(m1 *WebsocketProtocolMessage, m2 *WebsocketProtocolMessage) bool {
	if m1.MessageType != m2.MessageType {
		return false
	}

	if (len(m1.Parameters) == 0 && len(m2.Parameters) != 0) || (len(m1.Parameters) != 0 && len(m2.Parameters) == 0) {
		return false
	}

	if len(m1.Parameters) == 0 && len(m2.Parameters) == 0 {
		return true
	}

	for k, v := range m1.Parameters {
		if m2.Parameters[k] != v {
			return false
		}
	}

	for k, v := range m2.Parameters {
		if m1.Parameters[k] != v {
			return false
		}
	}

	return true
}

func testMessageIntegrity(t *testing.T, m *WebsocketProtocolMessage) {
	serialized := m.Serialize()

	parsed := ParseWebsocketProtocolMessage(serialized)

	if !compareMessages(m, parsed) {
		t.Error("Message integrity is not preserved. Message: " + m.Serialize() + " | Parsed: " + parsed.Serialize())
	}
}

func TestWebsocketProtocolMessage(t *testing.T) {
	// Test message integrity
	// meaning serialize -> deserialize does not change the message

	testMessageIntegrity(t, &WebsocketProtocolMessage{
		MessageType: "H",
	})

	testMessageIntegrity(t, &WebsocketProtocolMessage{
		MessageType: "PUSH",
		Parameters: map[string]string{
			"stream": "stream-id",
			"auth":   "auth-token-example",
		},
	})

	testMessageIntegrity(t, &WebsocketProtocolMessage{
		MessageType: "PULL",
		Parameters: map[string]string{
			"stream":                "stream-id",
			"auth":                  "auth-token-example",
			"only_source":           "true",
			"max_initial_fragments": "10",
		},
	})

	testMessageIntegrity(t, &WebsocketProtocolMessage{
		MessageType: "OK",
	})

	testMessageIntegrity(t, &WebsocketProtocolMessage{
		MessageType: "E",
		Parameters: map[string]string{
			"code":    "Error_Code",
			"message": "Example error message",
		},
	})

	testMessageIntegrity(t, &WebsocketProtocolMessage{
		MessageType: "F",
		Parameters: map[string]string{
			"duration": "2.5",
		},
	})

	testMessageIntegrity(t, &WebsocketProtocolMessage{
		MessageType: "CLOSE",
	})
}

File: client-js/tests/message-test.ts
// Test

"use strict";

import { CdnWebSocketMessage, serializeCdnWebSocketMessage, parseCdnWebSocketMessage } from "../src/message";

import assert from "assert";

function isEmptyMap(m: Map<string, string> | undefined): boolean {
    if (!m) {
        return true;
    }

    return m.size === 0;
}

function testMessage(msg: CdnWebSocketMessage) {
    const msgStr = serializeCdnWebSocketMessage(msg);
    const parsedMessage = parseCdnWebSocketMessage(msgStr);

    assert.equal(parsedMessage.type, msg.type);

    if ((isEmptyMap(parsedMessage.parameters) || isEmptyMap(msg.parameters)) && (!isEmptyMap(parsedMessage.parameters) || !isEmptyMap(msg.parameters))) {
        throw new Error("Parsed message parameters does not match with expected parameters. \n" + msgStr + "\n" + serializeCdnWebSocketMessage(parsedMessage))
    }

    if (isEmptyMap(parsedMessage.parameters) && isEmptyMap(msg.parameters)) {
        return true;
    }

    for (let [key, value] of msg.parameters?.entries() || []) {
        assert.equal(parsedMessage.parameters?.get(key), value);
    }

    for (let [key, value] of parsedMessage.parameters?.entries() || []) {
        assert.equal(msg.parameters?.get(key), value);
    }
}

describe("Message serialization / deserialization test", () => {
    it('Heartbeat message', async () => {
        testMessage({
            type: "H"
        });
    });

    it('Error message', async () => {
        testMessage({
            type: "E",
            parameters: new Map([
                ["code", "ERROR_CODE"],
                ["message", "Example error message"],
            ])
        });
    });

    it('PUSH message', async () => {
        testMessage({
            type: "PUSH",
            parameters: new Map([
                ["stream", "stream_id"],
                ["auth", "example-auth-token"],
            ])
        });
    });

    it('PULL message', async () => {
        testMessage({
            type: "PULL",
            parameters: new Map([
                ["stream", "stream_id"],
                ["auth", "example-auth-token"],
                ["only_source", "false"],
                ["max_initial_fragments", "10"]
            ])
        });
    });

    it('OK message', async () => {
        testMessage({
            type: "OK"
        });
    });

    it('Fragment message', async () => {
        testMessage({
            type: "F",
            parameters: new Map([
                ["duration", "1.5"],
            ])
        });
    });

    it('CLOSE message', async () => {
        testMessage({
            type: "CLOSE"
        });
    });
});
